### ISSUE TITLE
halogen-6: `Remove surface parameter` is a mistake

### ISSUE THREAD

**Issue by srghma** [purescript-halogen/purescript-halogen#686]
it was introduced in https://github.com/purescript-halogen/purescript-halogen/commit/6261ce8f57059774552d2bc626397891896da5e6

I think it's a mistake

the reasoning was that it is not being used 

but it is used

for example here https://github.com/boxyoman/purescript-halogen-reactnative/blob/2b0a3665756a2bc752558f4b9ac8958cf28d1438/src/Halogen/ReactNative/Driver.purs#L60

if it is removed - lot of code in purescript-halogen cannot be reused for implementing react-native renderer

basically I'm talking about all functions in src/Halogen/Component.purs file

here is the pr to return it https://github.com/srghma/purescript-halogen/pull/1/commits

**Comment by garyb** [comment]
Is anyone using `halogen-reactnative` though? I didn't even know it existed until now :wink:, plus it's unchanged in two years. I would assume it was more of an experiment than an ongoing concern.

**Comment by srghma** [comment]
Yes, but what's the reason to force creators of frameworks like this to have a copy of src/Halogen/Component.purs in their codebase with different HTML

P.S. maybe in future there will be native halogen framework that is using react native Java and iOS components but without react, who knows 🙂

**Comment by thomashoneyman** [comment]
As far as I can tell, the trade-off is between:

* Keeping the `surface` parameter, where the advantage is that Halogen components can render to something other than `HH.HTML`. For this to be a true advantage there has to be evidence that a) someone is doing this and b) that preserving the parameter in Halogen is less effort for users overall than requiring an alternate renderer to duplicate the `Component.purs` file.
* Dropping the `surface` parameter, where the advantage is that Halogen components have one less type parameter, which means one less thing for users to learn about and simpler type signatures for components. For this to be a true advantage there has to be evidence that folks find either this type in particular to be confusing or at least Halogen type signatures to be lengthy and confusing in general (as this is a very common type).

If I felt confident that the `surface` parameter was going to be used in a production setting by anyone to provide an alternate renderer then I would be in favor of keeping it. I don't think the two-year-old experiment with react-native is enough to set that expectation. (Unless I'm missing some potential implications for _testing_ Halogen components, which would change my mind).

I do feel confident, however, that folks regularly complain about Halogen having lengthy, hard-to-understand type signatures. I'm strongly in favor of changes that reduce type signatures with minimal impact on the library. If we can get away with explaining one less type parameter in a type that shows up everywhere in Halogen then I'm all for it.

> P.S. maybe in future there will be native halogen framework that is using react native Java and iOS components but without react, who knows 🙂

It's _possible_ that something like this might happen, but I'm not confident it will. Halogen has been the most popular UI library in PureScript for many years and the only attempt to do something like this is the react-native experiment you linked above. I just don't feel it's likely we are going to see major progress in this area. Even if we do, the effort of implementing something like what you've described is much larger than the penalty of copying `Component.purs`.

On the other hand, people will continue using regular Halogen, and removing the `surface` parameter can make the library a little nicer to use day-to-day.

**Comment by srghma** [comment]
> b) that preserving the parameter in Halogen is less effort for users overall

This is a breaking change for halogen libs

> Even if we do, the effort of implementing something like what you've described is much larger than the penalty of copying Component.purs.

Not only Component.purs I think

And also somehow keeping it in sync with halogen lib (using bash script that copies and `sed "s/HTML/YOURHTML/"` ?)

Anyway, do as you want

**Comment by thomashoneyman** [comment]
> This is a breaking change for halogen libs

Yes, it is a breaking change (albeit one that’s easy to fix via find / replace).

> And also somehow keeping it in sync with halogen lib

Yes, it would also have to be kept in sync. Fortunately this file doesn’t change much. But there’s no denying that removing the surface parameter makes implementing a react native renderer (for example) more difficult.

The question then is just whether keeping that use case easy is better than removing the parameter. It will be a while before Halogen 6 and if folks really want to implement another renderer then I at least am open to bringing it back. But without that I’m reluctant to add it.

> Anyway, do as you want

I’m not sure if you meant it this way, but this read a little brusque to me. If you feel strongly that this change shouldn’t have been made I am happy to keep discussing it. In the meantime I still prefer to keep the change.

**Comment by srghma** [comment]
> this read a little brusque

I mean I would keep the parameter because it would make my life maybe easier in future

I cannot argue against `less parameters -> easier to read` because that's subjective

**Comment by garyb** [comment]
I guess the way I look at it is this: Halogen has had the parametrised HTML for since v1.0, which was at the start of 2017. In that time, I've had two conversations with people explaining how to implement drivers as they were interested in trying something out, and there's the `halogen-reactnative` experiment. So 3 people that I'm aware of have actually tried to take advantage of this parameter, and I think for of two of them `HTML` still would have been fine as they were trying out a non-native `React` driver.

I've seen a lot more than 3 people make comments about the perceived difficulty/density of Halogen's types in that time. :wink:

If we ever get type synonyms to work as well as we'd like in PureScript, I'd have no reason at all to oppose it. At the moment if we were to make `type Component = Component' HH.HTML` or something, that helps with the readability of signatures, but will likely cause more difficulty with debugging errors for the people that have trouble with Halogen's types. If type synonyms were preserved, then this wouldn't be a problem as the error would refer back to `Component` again.

As it happens, preserving the parameter still has a bit of a problem when it comes to implementing other surfaces also: `GetRef` is hard coded to deal with `Element`s. There's no relationship between `surface` and `HalogenM`, and my attempts to make that work better have never been satisfactory, as it introduces quite a bit of new complexity (ideally you'd need something like type families to handle it well).

So basically, for now, I think doing anything we can that reduces the perceived complexity of Halogen is going to help its success more than preserving this one extension point that hasn't proven to be valuable in the 3.5 years it has been available. When/if synonyms work well, let's bring it back.

**Comment by srghma** [comment]
> GetRef is hard coded to deal with Elements. There's no relationship between surface and HalogenM

ok, I hear you, if this may result in faster code - I'm 100% behind

**Comment by thomashoneyman** [comment]
Feel free to continue discussing, but I'm going to close this issue as there's not a plan to act on it. Thanks!
